### PR TITLE
cdc/scheduler_v2: implement balancer

### PR DESCRIPTION
**Pull request by liuzix**
### What problem does this PR solve? 
- We need an interface responsible for balancing tables across the captures.

### What is changed and how it works?
- Defined the `balancer` interface.
- Implemented `tableNumberBalancer`.

### Check List 

Tests 

 - Unit test

### Release note 

```release-note
None
```

### CODE REVIEW

File: cdc/scheduler/balancer_test.go [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+
+	mockCaptureInfos := map[model.CaptureID]*model.CaptureInfo{
+		"capture-1": {
+			ID: "capture-1",
+		},
+		"capture-2": {
+			ID: "capture-2",
+		},
+		"capture-3": {
+			ID: "capture-2",
```
Comment (asddongmen) [inline cdc/scheduler/balancer_test.go:62]: capture-3?

File: cdc/scheduler/balancer_test.go [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
```
Comment (asddongmen) [inline cdc/scheduler/balancer_test.go:109]: ditto

**Comment by asddongmen** [comment]
rest LGTM

File: cdc/scheduler/balancer.go [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+	) (tablesToRemove []model.TableID)
+
+	// FindTarget returns a target capture to add a table to.
+	FindTarget(
+		tables *util.TableSet,
+		captures map[model.CaptureID]*model.CaptureInfo,
+	) (minLoadCapture model.CaptureID, ok bool)
+}
+
+// tableNumberBalancer implements a balance strategy based on the
```
Comment (zhaoxinyu) [inline cdc/scheduler/balancer.go:43]: Is it feasible to add a fine-grained balancer implementation? For example, we can rebalance tables based on their load such as the total number of events processed over a period of time

File: cdc/scheduler/balancer.go [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+// some victims are chosen and de-scheduled. Later, the victims
+// will be automatically rescheduled, during which the target captures
+// will be chosen so that the workload is the most balanced.
+type balancer interface {
+	// FindVictims returns a set of possible victim tables.
+	// Removing these tables will make the workload more balanced.
+	FindVictims(
+		tables *util.TableSet,
+		captures map[model.CaptureID]*model.CaptureInfo,
+	) (tablesToRemove []model.TableID)
```
Comment (nongfushanquan) [inline cdc/scheduler/balancer.go:34]: why not return map[CaptureID][tableIDList]

File: cdc/scheduler/balancer.go (hunk 3 above)
Comment (liuzix) [reply cdc/scheduler/balancer.go:43]: This design is not incompatible with a fine-grained design. If we can make the processors report their current workload more accurately, we will be able to find victims based on their actual workload.

File: cdc/scheduler/balancer.go (hunk 4 above)
Comment (liuzix) [reply cdc/scheduler/balancer.go:34]: We can return `[]TableRecord`. In the current design, we look up the `TableSet` to know which capture these tables are assigned to.

File: cdc/scheduler/balancer.go [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+	FindTarget(
+		tables *util.TableSet,
+		captures map[model.CaptureID]*model.CaptureInfo,
+	) (minLoadCapture model.CaptureID, ok bool)
+}
+
+// tableNumberBalancer implements a balance strategy based on the
+// current number of tables replicated by each capture.
+type tableNumberBalancer struct {
+	logger *zap.Logger
```
Comment (overvenus) [inline cdc/scheduler/balancer.go:46]: Why embed a logger?

Comment (liuzix) [reply cdc/scheduler/balancer.go:46]: This was intended for passing the `changefeedID` (and other relevant information for a larger scope).

**Review by nongfushanquan — APPROVED** [review]

**Review by asddongmen — APPROVED** [review]

File: cdc/scheduler/balancer.go [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+	// Algorithm overview: We try to remove some tables as the victims so that
+	// no captures are assigned more tables than the average workload measured in table number,
+	// modulo the necessary margin due to the fraction part of the average.
+	//
+	// In formula, we try to maintain the invariant:
+	//
+	// num(tables assigned to any capture) < num(tables) / num(captures) + 1
+
+	totalTableNum := len(tables.GetAllTables())
+	captureNum := len(captures)
```
Comment (amyangfei) [inline cdc/scheduler/balancer.go:108]: Could this be zero

**Review by amyangfei — APPROVED** [review]

**Pull request merged**